### PR TITLE
Improvements/refactors to substitution extraction routine

### PR DESCRIPTION
**Pull request by ehildenb**
Pulled out of: https://github.com/runtimeverification/k/pull/4625

This PR improves the substitution extraction machinery in `kast.manip`, and adds tests. This isn't used anywhere at the moment, but #4625 will start using it heavily.

- The code for `extract_substs` is simplified.
- The cases of circular substitutions are handled slightly more gracefully.

### CODE REVIEW

File: pyk/src/pyk/kast/manip.py [hunk 1]
```diff
@@ -215,48 +215,39 @@ def extract_rhs(term: KInner) -> KInner:
...
-                    and conjunct.args[1].label.name in {'_==K_', '_==Int_'}
-                ):
-                    subst = _subst_for_terms(conjunct.args[1].args[0], conjunct.args[1].args[1])
-
-                    if subst is not None:
-                        return subst
+    _subst = {}
+    rem_conjuncts: list[KInner] = []
 
+    def _extract_subst(_term1: KInner, _term2: KInner) -> tuple[str, KInner] | None:
```
Comment (tothtamas28) [inline pyk/src/pyk/kast/manip.py:221]: To reduce visual noise, consider:

```suggestion
    def _extract_subst(term1: KInner, term2: KInner) -> tuple[str, KInner] | None:
```

(Similarly for `_subst`.)

File: pyk/src/pyk/kast/manip.py [hunk 2]
```diff
@@ -215,48 +215,39 @@ def extract_rhs(term: KInner) -> KInner:
...
+        if (
+            (type(_term1) is KVariable and _term1.name not in _subst)
```
Comment (tothtamas28) [inline pyk/src/pyk/kast/manip.py:223]: This is considered more Pythonic:

```suggestion
            (isinstance(_term1, KVariable) and _term1.name not in _subst)
```

([PEP8](https://github.com/python/peps/blob/2ec2140ec83dcc6069e3e757641ca82809bc87eb/peps/pep-0008.rst?plain=1#L1455-L1468))

**Review by tothtamas28 — APPROVED** [review]

File: pyk/src/pyk/kast/manip.py (hunk 1 above)
Comment (ehildenb) [reply pyk/src/pyk/kast/manip.py:221]: Taken care of in https://github.com/runtimeverification/k/pull/4630/

File: pyk/src/pyk/kast/manip.py (hunk 2 above)
Comment (ehildenb) [reply pyk/src/pyk/kast/manip.py:223]: Taken care of in https://github.com/runtimeverification/k/pull/4630/

**Pull request merged**